File: src/engine/dhatu.rs
use deadpool_sqlite::Pool;
use serde_json::{json, Value};

pub async fn search_dhatu(query: &str, pool: &Pool) -> Vec<Value> {
    let q = query.to_string();
    pool.get().await.unwrap().interact(move |conn| {
        let mut res = Vec::new();
        let search_term = format!("%{}%", q);
        
        // This query finds the Root, Meaning, and generated forms (like 'Adara') just like the screenshot!
        let mut stmt = conn.prepare("
            SELECT p.dhatu_id, p.pratyaya, p.base_form, p.upasarga,
                   (SELECT value FROM info WHERE dhatu_id = p.dhatu_id AND key_name IN ('mUlaDAtuH', 'DAtuH') LIMIT 1) as root,
                   (SELECT value FROM info WHERE dhatu_id = p.dhatu_id AND key_name IN ('aTfaH', 'meaning', 'eng', 'hin') LIMIT 1) as meaning,
                   (SELECT value FROM info WHERE dhatu_id = p.dhatu_id AND key_name IN ('gaRaH', 'gana') LIMIT 1) as gana
            FROM participles p
            WHERE p.base_form LIKE ?1 OR root LIKE ?1 OR meaning LIKE ?1
            LIMIT 100
        ").unwrap();
        
        let mut rows = stmt.query([&search_term]).unwrap();
        while let Ok(Some(row)) = rows.next() {
            res.push(json!({
                "dhatu_id": row.get::<&str, String>("dhatu_id").unwrap_or_default(),
                "root": row.get::<&str, String>("root").unwrap_or_default(),
                "meaning": row.get::<&str, String>("meaning").unwrap_or_default(),
                "gana": row.get::<&str, String>("gana").unwrap_or_default(),
                "pratyaya": row.get::<&str, String>("pratyaya").unwrap_or_default(),
                "upasarga": row.get::<&str, String>("upasarga").unwrap_or_default(),
                "base_form": row.get::<&str, String>("base_form").unwrap_or_default()
            }));
        }
        res
    }).await.unwrap()
}

File: engine/dhatu.go
package engine

import (
	"database/sql"
	"fmt"
)

func hasTableDhatu(db *sql.DB, name string) bool {
	var n string
	err := db.QueryRow("SELECT name FROM sqlite_master WHERE type='table' AND name=?1", name).Scan(&n)
	return err == nil && n == name
}

// SearchDhatu queries Dhatu information and participles matching a search query
func SearchDhatu(db *sql.DB, query string) []map[string]any {
	var results []map[string]any
	searchTerm := fmt.Sprintf("%%%s%%", query)

	// new schema support
	if hasTableDhatu(db, "participle_forms") && !hasTableDhatu(db, "participles") {
		sqlQuery := `
			SELECT p.dhatu_id, p.variant, p.base, p.prefix,
			       COALESCE(i1.value, '') as root,
			       COALESCE(i2.value, '') as meaning,
			       COALESCE(i3.value, '') as gana
			FROM participle_forms p
			LEFT JOIN dhatu_info i1 ON i1.dhatu_id = p.dhatu_id AND i1.name IN ('OpadeSikasvarUpam','mUlaDAtuH','DAtuH')
			LEFT JOIN dhatu_info i2 ON i2.dhatu_id = p.dhatu_id AND i2.name IN ('arTaH','English Meaning','hindI arTa','aTfaH','meaning')
			LEFT JOIN dhatu_info i3 ON i3.dhatu_id = p.dhatu_id AND i3.name IN ('gaRaH','gana')
			WHERE p.base LIKE ?1 OR i1.value LIKE ?1 OR i2.value LIKE ?1
			LIMIT 100
		`
		rows, err := db.Query(sqlQuery, searchTerm)
		if err != nil {
			return results
		}
		defer rows.Close()
		for rows.Next() {
			var dhatuID, pratyaya, baseForm, upasarga, root, meaning, gana sql.NullString
			if err := rows.Scan(&dhatuID, &pratyaya, &baseForm, &upasarga, &root, &meaning, &gana); err != nil {
				continue
			}
			results = append(results, map[string]any{
				"dhatu_id":  dhatuID.String,
				"root":      root.String,
				"meaning":   meaning.String,
				"gana":      gana.String,
				"pratyaya":  pratyaya.String,
				"upasarga":  upasarga.String,
				"base_form": baseForm.String,
			})
		}
		return results
	}

	sqlQuery := `
		SELECT p.dhatu_id, p.pratyaya, p.base_form, p.upasarga,
		       COALESCE(i1.value, '') as root,
		       COALESCE(i2.value, '') as meaning,
		       COALESCE(i3.value, '') as gana
		FROM participles p
		LEFT JOIN info i1 ON i1.dhatu_id = p.dhatu_id AND i1.key_name IN ('mUlaDAtuH', 'DAtuH')
		LEFT JOIN info i2 ON i2.dhatu_id = p.dhatu_id AND i2.key_name IN ('aTfaH', 'meaning', 'eng', 'hin')
		LEFT JOIN info i3 ON i3.dhatu_id = p.dhatu_id AND i3.key_name IN ('gaRaH', 'gana')
		WHERE p.base_form LIKE ?1 OR i1.value LIKE ?1 OR i2.value LIKE ?1
		LIMIT 100
	`

	rows, err := db.Query(sqlQuery, searchTerm)
	if err != nil {
		return results
	}
	defer rows.Close()

	for rows.Next() {
		var dhatuID, pratyaya, baseForm, upasarga, root, meaning, gana sql.NullString
		if err := rows.Scan(&dhatuID, &pratyaya, &baseForm, &upasarga, &root, &meaning, &gana); err != nil {
			continue
		}

		results = append(results, map[string]any{
			"dhatu_id":  dhatuID.String,
			"root":      root.String,
			"meaning":   meaning.String,
			"gana":      gana.String,
			"pratyaya":  pratyaya.String,
			"upasarga":  upasarga.String,
			"base_form": baseForm.String,
		})
	}

	return results
}